src/main/java/org/datadog/jmxfetch/Instance.java

```diff
@@ -690,29 +860,61 @@ public List<String> getBeansScopes() {
      * Query and refresh the instance's list of beans. Limit the query scope when possible on
      * certain actions, and fallback if necessary.
      */
-    private void refreshBeansList() throws IOException {
-        this.beans = new HashSet<ObjectName>();
+    private void refreshBeansList(boolean isInitialQuery) throws IOException {
+        Set<ObjectName> newBeans = new HashSet<>();
         String action = appConfig.getAction();
         boolean limitQueryScopes =
                 !action.equals(AppConfig.ACTION_LIST_EVERYTHING)
                         && !action.equals(AppConfig.ACTION_LIST_NOT_MATCHING);
 
-        if (limitQueryScopes) {
-            try {
-                List<String> beanScopes = getBeansScopes();
-                for (String scope : beanScopes) {
-                    ObjectName name = new ObjectName(scope);
-                    this.beans.addAll(connection.queryNames(name));
+        this.beanAndAttributeLock.lock();
+        try {
+
+            boolean fullBeanQueryNeeded = false;
+            if (limitQueryScopes) {
+                try {
+                    List<String> beanScopes = getBeansScopes();
+                    for (String scope : beanScopes) {
+                        ObjectName name = new ObjectName(scope);
+                        newBeans.addAll(connection.queryNames(name));
+                    }
+                } catch (Exception e) {
+                    fullBeanQueryNeeded = true;
+                    log.error(
+                            "Unable to compute a common bean scope, querying all beans as a fallback",
+                            e);
                 }
-            } catch (Exception e) {
-                log.error(
-                        "Unable to compute a common bean scope, querying all beans as a fallback",
-                        e);
             }
-        }
 
-        this.beans = (this.beans.isEmpty()) ? connection.queryNames(null) : this.beans;
-        this.lastRefreshTime = System.currentTimeMillis();
+            if (fullBeanQueryNeeded) {
+                newBeans = connection.queryNames(null);
+            }
+            if (this.beanSubscriptionActive && !fullBeanQueryNeeded && !isInitialQuery) {
```

Comment:
TODO, decide how to enable this extra audit logging or if some other form of auditing would be more useful.